### DISCUSSION TITLE
Transaction API in Python binding, and more

### DISCUSSION THREAD

**Discussion by fantix** [geldata/gel#2841] [Ideas]
I ran into an issue in the Python binding with retrying transactions, and the topic grew a bit while I'm trying to fix the issue. So I'm starting a discussion here to collect some opinions.

## Background

The issue I met was actually pretty simple:

```python
async for tx in con.retrying_transaction():
    async with tx:  # <-- retryable error raised here
        await tx.execute("SELECT 42")
```

Theoretically the loop here will retry if we met a retryable error in the body. But in a special case when starting a transaction fails with a retryable error, the whole block fails without retrying. Because the transaction is started in `__aenter__()` when running `async with tx:`, while the recover logic is in `__aexit__()`. And `__aexit__()` doesn't capture errors raised in `__aenter__()`.

Because Python made sure that the `async with` block cannot be skipped from within `__aenter__()`, my solution is to make the retrying transaction start lazily until `tx.execute()` is called, after some discussion with @1st1. The draft PR is edgedb/edgedb-python#228, in short it will look like this:

```python
async for tx in con.retrying_transaction():
    async with tx:                     # doesn't do anything here
        await tx.execute("SELECT 42")  # actually start the transaction here and execute SELECT 42
        await tx.execute("SELECT 31")  # execute SELECT 31 only within the same transaction
```

However, more interesting topics came with it.

## Managed Transaction vs. Manual Transaction

To make the discussion simpler, we'll call the transaction controlled by a `with` or `async with` block a **Managed Transaction**, for example:

```python
async for tx in con.retrying_transaction():
    async with tx:
        # This is a managed transaction.
        # The transaction will be automatically comitted
        # when exiting the `async with` block without an exception,
        # or rolled back if an error occured.
```

Meanwhile, if a transaction is manually controlled by `tx.start()`, `tx.commit()` or `tx.rollback()`, we call it a **Manual Transaction**, like this:

```python
tx = await con.raw_transaction()
await tx.start()
try:
    await tx.execute("SELECT 42")
except Exception:
    await tx.rollback()
else:
    await tx.commit()
```

## Explicit Transaction Start

Now going back to our topic - we will make `retrying_transaction()` start lazily. Before this PR, manual transaction is not allowed in `retrying_transaction()` because the retrying logic requires a Python context to capture the retryable errors, and enabling retrying logic for manual transaction is a terrible developer experience. (Imagine that you need to capture errors manually, rollback and raise non retryable errors.)

Now the transaction will not start as we enter the `async with tx:` block. Can we allow `tx.start()` in a managed transaction as a measure to explicitly start the lazy transaction? For example:

```python
async for tx in con.retrying_transaction():
    async with tx:                     # doesn't do anything here
        await tx.start()               # actually start the transaction here explicitly
        await tx.execute("SELECT 42")  # execute SELECT 42 within the transaction
    # When exiting the `async with` block, still follow the rules of managed transactions
```

This mitigates a little bit of the issue of implicit transaction start, while the downside is possibly introducing confusion between managed and manual transaction by using methods from both sides in one place.

## More about Explicitness

A small side-topic we discussed was, should an empty lazy transaction stay lazy? Like:

```python
async for tx in con.retrying_transaction():
    async with tx:                     # doesn't do anything here
        await asyncio.sleep(1)
    # exiting the `async with` block - what to do here?
```

It's natual that we simply do nothing exiting a lazy transaction that never started. On the other hand, `async with tx:` does try to start a transaction before this PR, no matter what is in its body. Do we want to keep this behavior after the PR and maintain some level of explicitness? The answer is likely "no" because we'll advertise the lazy transaction as a feature.

## Laziness in Retrying Only

Because for now the API (and underlying source code) shares the same signature between `retrying_transaction()` and `raw_transaction()`, so introducing laziness in one may confuse the user of the other. However, we would like to keep `raw_transaction()` as strictly "raw", so that there is no fancy magic like laziness.

## What about `rollback()`

@1st1 mentioned a valid use case when we actually wanted to eventually roll back a transaction:

```python
async for tx in con.retrying_transaction():
  async with tx:
     report = await calculate_report_maybe_with_some_inserts(tx)
     await tx.rollback()
```

This is not working both before and after the PR, but should potentially be supported in some way. Like mentioned previsouly, using `rollback()` or even `commit()` in a managed transaction is really confusing, we should even think about removing them (?). So @1st1 suggested one possible solution:

```python
async for tx in con.retrying_transaction(rollback=True):
    async with tx:
        report = await calculate_report_maybe_with_some_inserts(tx)
```

I have another (borrowed from [GINO](https://python-gino.org/docs/en/1.0/how-to/transaction.html#basic-usage)):

```python
async for tx in con.retrying_transaction():
    async with tx:
        report = await calculate_report_maybe_with_some_inserts(tx)
        tx.raise_rollback()
```

There could potentially be more solutions, but I'll stop here for discussion.

## The Must-Nots

However, we are certain about a few cases that are illegal:

```python
async for tx in con.retrying_transaction():
    await tx.execute("SELECT 31")  # must not start a transaction implicitly
```

```python
async for tx in con.retrying_transaction():
    async with tx:
        await tx.commit()  # mixing manual control in managed transaction
```

```python
async for tx in con.retrying_transaction():
    await tx.commit()  # no transaction to commit
```

```python
async for tx in con.retrying_transaction():
    await tx.start()  # manual transaction in `retrying_transaction()`
```

## Other Language Bindings

We should at least make sure that:

1. The retrying issue with `start()` doesn't exist or should be fixed.
2. The `retrying_transaction()` and `start()`, `rollback()` and `commit()` should have consistent semantic meaning across all bindings, if laziness is somehow implemented differently (the root cause is Python-only).

**Comment by 1st1** [comment]
Curious to hear arguments from other people, but right now I'm thinking that managed transactions are already quite complex, so allowing any of the following:

```python
async for tx in con.retrying_transaction():
  async with tx:
     await tx.start()
     # or
     await tx.commit()
     # or
     await tx.rollback()
```

... doesn't make much sense to me. IMO `async for tx in con.retrying_transaction()` and `async for tx in con.retrying_transaction(rollback=True)` cover pretty much all scenarios in a fairly unambiguous way. If we add these methods, then there would be some gotchas associated with them:

* can I have some code after I manually `commit()`?
* can I have some code call `tx.query()` after I call `tx.commit()`? The answer is no, obviously.
* etc.

So I'd keep it simple.

As for lazy transactions and "empty" with blocks, as in:

```python
async for tx in con.retrying_transaction():
    async with tx:                     # doesn't do anything here
        await asyncio.sleep(1)
    # exiting the `async with` block - what to do here?
```

IMO we shouldn't open a transaction just to immediately close it. It's pointless. On the contrary, I think we should detect that no query was made in that transaction and issue a warning.

**Comment by tailhook** [comment]
I'm thinking that we can capture error in `__enter__` and throw it on the next `query` or directly handle in `__exit__` if nothing was called. This is only suboptimal if user has some external resource access at the start of atransaction before any query (but it's even more confusing in lazy variant because starting transaction is semantically meaningful, and user need to know to put `.start()` in this case).

**Comment by elprans** [comment]
> starting transaction is semantically meaningful

I don't think a lazily-started transaction is any more magical than an auto-retrying one. In both cases you can't know the precise timing of transaction start and end. And that's OK! Transactions only guarantee db operation consistency inside the block.

**Reply by tailhook** [reply]
Something like this may be problem:
```
with tx:
   fetch("some.external.resource")
   tx.query(...)
```
You expect start to happen before `fetch` (i.e. postgres starting a snapshot at that time?). 

I'm not sure if there are many practical cases where it matters. But as I have already said, explaining this distinction to user is very unfortunate.

**Reply by elprans** [reply]
I don't understand.  What's the issue here?  You are either running queries in a transaction or you aren't. Where is the distinction and what needs explaining here?

**Comment by elprans** [comment]
> we will make `retrying_transaction()` start lazily

I think this is a good approach.  In fact, PostgreSQL itself implements transactions lazily.  `START TRANSACTION` does nothing special aside from setting a flag (`CurrentTransactionState->blockState = TBLOCK_BEGIN`).  The actual heavy work of allocating transaction resources happens on the first meaningful command.

> Can we allow `tx.start()` in a managed transaction as a measure to explicitly start the lazy transaction?

There's nothing special about starting a transaction block by itself.  The only semantic meaning is that all _database_ operations must happen in a transaction.  Allowing an explicit `tx.start()` inside a `with tx` block would be very weird.  

Furthermore, I don't think we should allow an explicit `commit()` or `rollback()` inside a `with` block either.  Both are just as harmful and confusing as an explicit `start()`.  There should be only one visible transaction boundary: the `with` block itself.  Consider the following code:

```
def some_function(tx):
    if condition:
        tx.rollback()

async with tx:
   await tx.query(...)
   ...
   some_function(tx)
   await tx.query(...) # <<< Is this inside a transaction or not? 
                       # as a reader, I would assume it to always be, 
                       # but I'd be wrong!
```

Yes, we would raise an error if you try to `query` on a closed transaction object, but I don't think it's a good design at all.  If you need to rollback manually, just raise an exception.

> I think we should detect that no query was made in that transaction and issue a warning.

I don't think such warning is necessary.  Because transactions are lazy, you're not even wasting any resources, it's almost free.  I think we should use warnings sparingly, and only in cases that are actually harmful or costly, otherwise we risk creating an impression of a "warning-happy" library and causing developers to ignore the warnings that actually matter.

**Reply by 1st1** [reply]
what about my idea of adding a `rollback` arg?

**Reply by tailhook** [reply]
>  In fact, PostgreSQL itself implements transactions lazily.

I remember you saying otherwise when we were discussing transaction conflict and deadlock errors.

**Comment by fmoor** [comment]
@elprans what about `raw_transaction()`? It seems like the same reasoning about `start()`, `commit()` and `rollback()` would apply here too.

**Comment by 1st1** [comment]
raw_transaction must retain explicit methods, because it has to be fully composable so that it can be used to build higher level primitives on top of it.

**Reply by elprans** [reply]
`raw_transaction()` can have explicit methods, but I think that even in this case a context manager flavor of it should return an object _without_ explicit methods, for basically the same reasons as the managed transaction.  So the following is legal:

```python
tx = conn.raw_transaction()
tx.start()
tx.query()
tx.commit()
```

but the this is not:

```python
with conn.raw_transaction() as tx:
    tx.query()
    if foo:
        tx.rollback()
```
